master/pkg/searcher/asha.go

```diff
@@ -239,12 +239,36 @@ func (s *asyncHalvingSearch) progress(float64) float64 {
 	if allTrials == s.maxTrials {
 		progress = math.Max(float64(s.trialsCompleted)/float64(s.maxTrials), progress)
 	}
+	// Cap progress at 95% for InvalidHP cases
+	if progress > 1 {
+		progress = float64(0.95)
+	}
 	return progress
 }
 
 func (s *asyncHalvingSearch) trialExitedEarly(
 	ctx context, requestID RequestID, exitedReason workload.ExitedReason,
 ) ([]Operation, error) {
+	if exitedReason == workload.InvalidHP {
+		var ops []Operation
+		s.earlyExitTrials[requestID] = true
+		ops = append(ops, NewClose(requestID))
+		s.closedTrials[requestID] = true
+		// Remove metrics associated with InvalidHP trial across all rungs
+		highestRungIndex := s.trialRungs[requestID]
+		for rungIndex := 0; rungIndex <= highestRungIndex; rungIndex++ {
+			rung := s.rungs[rungIndex]
+			rung.metrics = nil
```

Comment:
blocking: here don't we need to iterate over `rung.metrics` and just delete the metric where `metric.requestID == requestID`, not just nil out the metrics for the whole rung, right?